### PR TITLE
--character-mapping option

### PR DESCRIPTION
**Pull request by irskep**
This change allows you to write a JSON file like this:

```json
{
    "aircraft": "a",
    "city-planning": "\uf001"
}
```

If an SVG file name matches a key, then it will be assigned to the code point in the value.

This argument is optional. It should have no effect on existing workflows.

If a character mapping is specified and any SVG file is not present in the mapping, it will be reported at the end of the run.

### CODE REVIEW

File: icons2font.py [hunk 1]
```diff
@@ -218,60 +249,6 @@ def compile_path(commands):
             buf.append(str(n))
     return " ".join(buf)
 
-def compute_minrec():
```
Comment (irskep) [inline icons2font.py]: This function was unused.

File: icons2font.py [hunk 2]
```diff
@@ -115,6 +110,42 @@
...
 
+
+class GlyphNameMapper(object):
+    def __init__(self, character_mapping):
+        super(GlyphNameMapper, self).__init__()
+        self.character_mapping = character_mapping
+
+        self.used_values = set(self.character_mapping.viewvalues())
+        self.new_mappings = OrderedDict()
+        self.value = ord(max(self.used_values)) if len(self.used_values) else USER_AREA
```
Comment (scottrogowski) [inline icons2font.py]: could we call self.value something different? Or, alternatively, could we put a short blurb at the top of this class that explains what it does?

File: icons2font.py [hunk 3]
```diff
@@ -115,6 +110,42 @@
...
+
+    def advance_to_unused_value(self):
+        while self.value in self.used_values:
+            self.advance()
+
+    def get_glyph_name(self, friendly_name):
+        if friendly_name in self.character_mapping:
+            return self.character_mapping[friendly_name]
+        else:
+            this_name = self.value
```
Comment (scottrogowski) [inline icons2font.py]: why create a new variable here?

File: icons2font.py [hunk 4]
```diff
@@ -115,6 +110,42 @@
...
+            self.advance_to_unused_value()
+            self.new_mappings[friendly_name] = unichr(this_name)
+            return self.new_mappings[friendly_name]
+
+    def log_new_mappings_if_necessary(self):
+        if len(self.character_mapping) and len(self.new_mappings):
+            log.info("Character mapping was missing some characters. Suggested additions:\n{}".format(
+                '\n'.join([
+                    '"{}": "\\u{:x}",'.format(name, ord(char))
+                    for name, char in self.new_mappings.viewitems()
```
Comment (scottrogowski) [inline icons2font.py]: is it possible to divide this up?

File: icons2font.py (hunk 2 above)
Comment (irskep) [inline icons2font.py]: `next_glyph_name`?

File: icons2font.py (hunk 3 above)
Comment (irskep) [inline icons2font.py]: Why not? Also, `self.value` gets mutated a couple lines later and we need to hang on to this value.

File: icons2font.py (hunk 4 above)
Comment (scottrogowski) [inline icons2font.py]: til about viewitems

Comment (irskep) [inline icons2font.py]: Sure it's possible, it just didn't seem helpful. This way the format strings appear in the same order that they appear in the output.

File: examples/nounfont-character-mapping.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+    "aircraft": "a",
+    "city-planning": "\uf001"
+}
```
Comment (scottrogowski) [inline examples/nounfont-character-mapping.json]: hate to be an ass but newline please. It's helpful to install a linter to do this. Ignoring these and trailing whitespace makes future diffs harder to review

Comment (phil303) [inline examples/nounfont-character-mapping.json]: not being an ass

Comment (irskep) [inline examples/nounfont-character-mapping.json]: No problem at all! I just fixed my Sublime config.

File: icons2font.py (hunk 4 above)
Comment (scottrogowski) [inline icons2font.py]: how about

    suggested_additions = '\n'.join(['"{}": "\\u{:x}",'.format(name, ord(char))
                                                      for name, char in self.new_mappings.viewitems()]
    log.info("Character mapping was missing some characters. Suggested additions:\n{}".format(suggested_additions))

just makes it easier to read

Comment (irskep) [inline icons2font.py]: :+1:

File: icons2font.py (hunk 2 above)
Comment (scottrogowski) [inline icons2font.py]: could work. Might be a misnomer since it's an integer. I'm fine if that's what sounds best though

Comment (irskep) [inline icons2font.py]: Yeah, this script is screwy with names. My working vocabulary is:

* glyph - the idea and the visual representation of it (both airplane, and drawing-of-airplane)
* Friendly name - name of the SVG file without the extension
* glyph name - integer representing a Unicode code point to which a glyph will be assigned
* character - glyph name converted to a str/char data type

File: icons2font.py (hunk 3 above)
Comment (scottrogowski) [inline icons2font.py]: didn't see that before. Got it

File: examples/nounfont-character-mapping.json (hunk 5 above)
Comment (irskep) [inline examples/nounfont-character-mapping.json]: (fixed)

**Comment by scottrogowski** [comment]
lgtm. Merge at will

**Pull request merged**